exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+	clusterList := &clusterv1.ClusterList{}
+	if err := r.Client.List(ctx, clusterList, client.InNamespace(clusterResourceSet.Namespace), client.MatchingLabels(clusterResourceSet.Spec.ClusterSelector.MatchLabels)); err != nil {
+		return nil, errors.Wrap(err, "failed to list clusters")
+	}
+
+	clusters := []*clusterv1.Cluster{}
+	for i := range clusterList.Items {
+		m := &clusterList.Items[i]
+		if m.DeletionTimestamp.IsZero() {
+			clusters = append(clusters, m)
+		}
+	}
+	return clusters, nil
+}
+
+func (r *ClusterResourceSetReconciler) ApplyClusterResourceSet(ctx context.Context, cluster *clusterv1.Cluster, clusterResourceSet *addonsv1.ClusterResourceSet) error {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	logger.Info("Applying ClusterResourceSet to cluster", "Cluster", cluster.Name)
+
+	err := r.applyClusterResourceSet(ctx, cluster, clusterResourceSet)
+	if err != nil {
+		logger.Error(err, "Failed to apply resources")
+		return err
+	}
+	conditions.MarkTrue(clusterResourceSet, addonsv1.ResourcesAppliedCondition)
+
+	return nil
+}
+
+// applyClusterResourceSet applies resources in a ClusterResourceSet to a Cluster. Once applied, a record will be added to the
+// cluster's ClusterResourceSetBinding.
+// In ApplyOnce strategy, resources are applied only once to a particular cluster. ClusterResourceSetBinding is used to check if a resource is applied before.
+// It applies resources best effort and continue on scenarios like: unsupported resource types, failure during creation, missing resources.
+// TODO: If a resource already exists in the cluster but not applied by ClusterResourceSet, the resource will be updated ?
+func (r *ClusterResourceSetReconciler) applyClusterResourceSet(ctx context.Context, cluster *clusterv1.Cluster, clusterResourceSet *addonsv1.ClusterResourceSet) error {
+	logger := r.Log.WithValues("ClusterResourceSet", clusterResourceSet.Name, "namespace", clusterResourceSet.Namespace)
+
+	c, err := remote.NewClusterClient(context.Background(), r.Client, client.ObjectKey{Namespace: cluster.Namespace, Name: cluster.Name}, r.scheme)
+	if err != nil {
+		return err
+	}
+
+	// Get ClusterResourceSetBinding object for the cluster.
+	clusterResourceSetBinding, err := GetorCreateClusterResourceSetBinding(ctx, r.Client, cluster, clusterResourceSet)
+	if err != nil {
+		return err
+	}
+
+	// Initialize the patch helper.
+	patchHelper, err := patch.NewHelper(clusterResourceSetBinding, r.Client)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		// Always attempt to Patch the ClusterResourceSetBinding object after each reconciliation.
+		if err := patchHelper.Patch(ctx, clusterResourceSetBinding); err != nil {
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	InitClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
```

Comment:
Would it make sense for GetorCreateClusterResourceSetBinding to do this?

Reply:
There is a deferred patching here, so need to create patch before making any changes on clusterResourceSetBinding.

Reply:
Same comment below. Resolving this one.